### ISSUE TITLE
Rust unable to optimize away assert vs. clang. 

### ISSUE THREAD

**Issue by jrmuizel** [rust-lang/rust#93036]
```rust
pub fn f(arr: &mut [u32]) {
    for i in 0..arr.len() {
        for j in 0..i {
            assert!(j < arr.len());
        }
    }
}
```
Rustc does not optimize away the the assert. Adding `-C passes=constraint-elimination` doesn't help.

However, clang does optimize away this assert when compiled with `-O2 -mllvm -enable-constraint-elimination`
```c
void f(int *ptr, size_t len) {
    for (size_t i = 0; i < len; i++) {
            for (size_t j = 0; j < i; j++) {
                assert(j < len);
            }
    }
}
```

GCC is also able to remove the assert with GCC trunk.

**Comment by dalcde** [comment]
More generally, rust/llvm does not understand that inequality is transitive. The following code generates three inequality asserts in the assembly:
```rust
pub fn cmp(a: u32, b: u32, c: u32) {
    assert!(a < b);
    assert!(b < c);
    assert!(a < c);
}
```
Rust playground: https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=8cfdb982df6b7e781b398b9408b3cd10

**Comment by jrmuizel** [comment]
> More generally, rust/llvm does not understand that inequality is transitive. The following code generates three inequality asserts in the assembly:

Adding `-C passes=constraint-elimination` fixes that.

**Comment by jrmuizel** [comment]
but not if the types are changed to `i32` for which I filed: https://github.com/llvm/llvm-project/issues/53273

**Comment by the8472** [comment]
Assuming you need this for slicing then creating a temporary works.

```rust
pub fn f(arr: &mut [u32]) {
    for i in 0..arr.len() {
        let subslice = &arr[0..i];
        for j in 0..subslice.len() {
            assert!(j < subslice.len());
        }
    }
}
```

https://rust.godbolt.org/z/n6sa6Gf34

**Comment by nikic** [comment]
This still requires ConstraintElimination to be enabled on the LLVM side. If it is, then this will land with the LLVM 15 upgrade in ~August.

**Comment by andjo403** [comment]
this is fixed in latest stable rustc 1.73 see https://rust.godbolt.org/z/G1E35e375
